Src/Plugins/Security/SEC101_026.MailgunApiKeyValidator.cs

```diff
@@ -0,0 +1,118 @@
+                                                ref message);
+        }
+
+        protected override string IsValidStaticHelper(ref string matchedPattern,
+                                                      ref Dictionary<string, string> groups,
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("key", out string key) ||
+                !groups.TryGetNonEmptyValue("account", out string account))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint
+            {
+                Key = key,
+                Account = account,
+            }.ToString();
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText,
+                                                       ref string message)
+        {
+            var fingerprint = new Fingerprint(fingerprintText);
+
+            string key = fingerprint.Key;
+            string account = fingerprint.Account;
+
+            try
+            {
+                using HttpClient client = CreateHttpClient();
+
+                string credentials = $"api:{key}";
+                byte[] bytes = Encoding.UTF8.GetBytes(credentials);
+                credentials = Convert.ToBase64String(bytes);
+
+                client.DefaultRequestHeaders.Authorization =
+                    new AuthenticationHeaderValue("Basic", credentials);
+
+                var content = new MultipartFormDataContent();
+                content.Add(new StringContent(Guid.NewGuid().ToString()), "subject");
+
+                using HttpResponseMessage response = client
+                    .PostAsync($"https://api.mailgun.net/v3/{account}/messages", content)
+                    .GetAwaiter()
+                    .GetResult();
+
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.BadRequest:
+                    {
+                        return ReturnAuthorizedAccess(ref message, asset: account);
+                    }
+
+                    case HttpStatusCode.Unauthorized:
+                    {
+                        return ReturnUnauthorizedAccess(ref message, asset: account);
+                    }
+
+                    default:
+                    {
+                        message = $"An unexpected response was returned : '{response.StatusCode}'";
+                        break;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                return ReturnUnhandledException(ref message, e);
+            }
+
+            return ReturnUnknownAuthorization(ref message, asset: account);
```

Comment:
>return ReturnUnknownAuthorization(ref message, asset: account); [](start = 12, length = 63)

if the status code is NotFound, for example. We would change the message in line 105. but, when we get  here, we would change again.
With that, i would move this and return nameof(ValidationState.Unknown) in line 106 instead of the break.

Reply:
we should return unknown at 106 and make this an exceptional path (as a non-feasible code path) or do what I did, which is to construct a message on line 105 and now simply return unknown on line 115. what do you think?

---
In reply to: [575899988](https://github.com/microsoft/sarif-pattern-matcher/pull/210#discussion_r575899988) [](ancestors = 575899988)